### PR TITLE
Feature Request: Persistent script

### PR DESCRIPTION
**Pull request by umireon**
As the author of the Emmet plugin for CotEditor,
I want persistent and preloaded execution mode to be integrated in CotEditor
To make the plugin works with state and breezingly fast.

The thing I want to be introduced is `PersistentOSAScript` in 4d841f4.
In this mode, the instance of `OSAScript` is initiated only once and execute the script without spawning the separate process as `NSUserAppleScriptTask` does.

The execution model can be specified with the key `CotEditorExecutionModel ` in `Contents/Info.plist` as [scripting hook](https://github.com/coteditor/CotEditor/pull/649).

```plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>CotEditorExecutionModel</key>
  <string>persistent</string>
</dict>
</plist>
```

### Add persistent mode for AppleScript scripts (4d841f4)

`NSUserAppleScriptTask` is the recommended way to execute user-supplied
scripts and the only way to execute them outside of the sandbox.
However, `NSUserAppleScriptTask` has a performance issue because it
compiles a script and spawns a separate process every time the script
would be executed.

This change adds the persistent execution mode with `OSAScript`, which
compiles scripts only once but executes them inside the sandbox.

### CODE REVIEW

**Comment by 1024jp** [comment]
Thanks for the pull request.

However, I feel that your design is a bit too complex for a text editor like CotEditor. I mean about switching two different modes: the persistent one and the normal one. Give me a time. I'll consider a bit. I have to be careful to adding new features that might be used rarely.

Secondary, could you separate pull request into multiple requests so that I can review each refactoring/feature?

**Comment by 1024jp** [comment]
I also wondering if loading all scripts to memory is good way.
I suppose 99% of users actually don't use scripts. Therefore I'd like to avoid reading a lot of files at application launch.

File: CotEditor/Sources/Script.swift [hunk 1]
```diff
@@ -27,92 +27,215 @@
...
+    // MARK: Methods
+    
+    /// Execute the script with the default way.
+    func run() throws
+    
+    
+    /// Execute the script by sending it the given Apple event.
+    ///
+    /// Events the script cannot handle must be ignored with no errors.
+    func run(withAppleEvent event: NSAppleEventDescriptor?) throws
```
Comment (1024jp) [inline CotEditor/Sources/Script.swift]: As long as `Script` protocol is also used with non-AppleScript objects, it shouldn't have AppleEvent specific methods.

**Comment by umireon** [comment]
Thank you for reviewing, @1024jp !
I'm sorry mixing up this PR and created refactoring PR at #657

**Comment by umireon** [comment]
@1024jp 

> I also wondering if loading all scripts to memory is good way.

I meant only scripts in persistent mode will be preloaded.
Any other scripts using `NSUserScriptTask` will never be loaded at startup and in fact preloading them is technically impossible.

**Comment by 1024jp** [comment]
>> I also wondering if loading all scripts to memory is good way.
> I meant only scripts in persistent mode will be preloaded.

Sorry, it was my misunderstanding. I realized that just now reading codes more carefully.

**Comment by umireon** [comment]
This PR will be rebased and force-pushed in favor of #657.

**Comment by umireon** [comment]
This feature enable scripts to implement fancy caret motions.

screencast 2016-12-11 20-27-31 mp4

File: CotEditor/Sources/Script.swift [hunk 2]
```diff
@@ -187,7 +208,7 @@ extension Script {
 
 // MARK: -
 
-final class AppleScript: Script {
+final class UnrestrictedAppleScript: Script {
```
Comment (1024jp) [inline CotEditor/Sources/Script.swift]: IMO, `AppleScript` class doesn't need to be renamed.

**Comment by 1024jp** [comment]
And after merging, could you also write a help page shortly for this feature?

**Comment by 1024jp** [comment]
It seems you've updated the code in accordance with my suggestion.
Can I merge it now?

**Comment by umireon** [comment]
@1024jp Apologies for latency.

**TL; DR** I had doubted the usefulness of this PR and revealed that this PR is effective in certain situations such as JXA scripts requiring persistency.

I have been inspecting the behavior difference between AppleScript and JXA in the `persistent` mode.
I was surprised by the fact that an AppleScript script stored in a `scpt` file has the capability to make its global variable persist by holding the value in the `scpt` file even with the sandboxed application.
There still be no such easy ways with JXA and therefore the purpose of this PR is to help script writers in JXA.

If this fact, which means the target of this feature is very minor, is acceptable, this PR is ready to merge 😄 and I will be happy to write the new help page ❓.
If not, it's OK 👍. I still have options to implement the fancy caret motions and such thingy, with somewhat cumbersome 😥 hacks (i.e. call `NSUserScriptTask` from AppleScript or use an external storage).

**Conclusion** I hope this PR would be merged: This is a very reliable way for JXA script writers to implement the persistency and I also want to write a vim keybinding plugin and this will need the proposed persistency mode for the performance reason.

PS I am sorry for confusing you due to lacking for the domain knowledge on AppleScript and macOS Automation.

Thanks

**Comment by 1024jp** [comment]
I'm planning to submit CotEditor 3.1.3 within few days. I'm just wondering if you @umireon is going to finish the help contents before that.
You don't need to hurry. If it takes time, I'll postpone the help contents part to the next version (3.1.4 or 3.2.0).

**Comment by umireon** [comment]
@1024jp 
I'm sorry, I had been busy with housekeeping tasks at the end of last year and forget to work on this 🙇🙇🙇.
Fortunately, I can have time to write them in this weekend, and I will submit the PR in the next two days.

**Comment by 1024jp** [comment]
Thanks for the reply. np at all. Then, I'll release CotEditor 3.1.3 in the next week after your PR.
It seems I also need to work for the issue #669 before the release of v3.1.3🤔.

**Comment by 1024jp** [comment]
@umireon CotEditor 3.1.3 は App Store に回しちゃったので、ヘルプはそれ以降のバージョンで加えましょう。3.1.3 ではリリースノートには加えていませんが、この PR 自体は生きています。ので、隠し機能ですね。
